### PR TITLE
fix #108495, postfix decrement and prefix decrement has no warning

### LINKED ISSUE #108495: postfix decrement and prefix decrement has no warning
### Code

```Rust
fn main(){
    let mut i=10;
    while i !=0{
        i--;
    }
}
```

### Current output

```Shell
error: expected expression, found `;`
 --> teste.rs:4:12
  |
4 |         i--;
  |            ^ expected expression

error: aborting due to previous error
```

### Desired output

```Shell
error: Rust has no postfix decrement operator
 --> ./testt.rs:4:10
  |
4 |         i--;
  |          ^^ not a valid postfix operator
  |
help: use `-= 1` instead
  |
4 |         i -= 1;
  |           ~~~~

error: aborting due to previous error
```

### Rationale and extra context

rust compiler throws error when postfix increment of prefix increment is used, but when decrement is it does not show where the problem is 

 for example this just compiles with no warning
```rust
fn main(){
    let i=10;
    let c=--i +1;
    println!("{c}");
}
```
the output will be 11 which is not as expected

edit
--i is double negation is rust which I didn't know
but i-- give wrong error

### Other cases

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by nx2k3**
Fixes #108495

### CODE REVIEW

**Review by Noratrieb — CHANGES_REQUESTED** [review]
Thank you for the PR, but we cannot do this recovery this way.

```rust
let _ = i--i;
```
and
```rust
let _ = --i;
```
compile today (double negation). The only case we would be able to handle is the case of something like `i--+i`, which is not allowed today. You'd have to be careful to make sure that whatever comes after the second `-` is not allowed to start an expression (there should be a helper function for that). I _think_ it should work then, but I'm not entirely sure.

I see your issues mentions a "warning". I think emitting a warning on double negation is a great idea!

**Comment by fmease** [comment]
> I see your issues mentions a "warning". I think emitting a warning on double negation is a great idea!

Fyi, Clippy already emits a warning for this by default:

```
warning: `--x` could be misinterpreted as pre-decrement by C programmers, is usually a no-op
 --> e.rs:3:9
  |
3 | let _ = --i;
  |         ^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#double_neg
  = note: `#[warn(clippy::double_neg)]` on by default
```

Not sure if it's worth moving this Clippy lint into rustc (this sometimes does happen though, although I don't know what the process is usually like).

**Comment by WaffleLapkin** [comment]
> this sometimes does happen though, although I don't know what the process is usually like

The process is more or less "ask T-compiler for approval, if they give a green light, make a PR that adds (to rustc) and deprecates (in clippy) the lint". See https://github.com/rust-lang/rust/issues/99272 and https://github.com/rust-lang/rust/pull/99696 as an example.

Although I don't think it's worth it in the case of `clippy::double_neg` lint.

File: compiler/rustc_parse/src/parser/expr.rs [hunk 1]
```diff
@@ -550,10 +561,10 @@ impl<'a> Parser<'a> {
...
             // `~expr`
             token::Tilde => make_it!(this, attrs, |this, _| this.recover_tilde_expr(lo)),
-            // `-expr`
-            token::BinOp(token::Minus) => {
-                make_it!(this, attrs, |this, _| this.parse_unary_expr(lo, UnOp::Neg))
-            }
+            // // `-expr`
+            // token::BinOp(token::Minus) => {
+            //     make_it!(this, attrs, |this, _| this.parse_unary_expr(lo, UnOp::Neg))
+            // }
```
Comment (WaffleLapkin) [inline compiler/rustc_parse/src/parser/expr.rs:567]: Did you mean to remove this? (given the comment below, I don't think this should be touched at all...)

File: compiler/rustc_parse/src/parser/expr.rs [hunk 2]
```diff
@@ -595,6 +606,27 @@ impl<'a> Parser<'a> {
                 let operand_expr = this.parse_dot_or_call_expr(Default::default())?;
                 this.recover_from_prefix_increment(operand_expr, pre_span, starts_stmt)
             }
+            // Recover from `--x`:
+            token::BinOp(token::Minus)
+                if this.look_ahead(1, |t| *t == token::BinOp(token::Minus))
+                    && !this.token.can_begin_expr() =>
```
Comment (WaffleLapkin) [inline compiler/rustc_parse/src/parser/expr.rs:612]: This is a part of a `match` on `this.token.uninterpolate().kind` ⇒ `this.token` is always `token::BinOp(token::Minus)` ⇒ `!this.token.can_begin_expr()` is always `false` ⇒ this is a dead branch.

**Review by WaffleLapkin — CHANGES_REQUESTED** [review]

File: compiler/rustc_parse/src/parser/expr.rs (hunk 2 above)
Comment (nx2k3) [reply compiler/rustc_parse/src/parser/expr.rs:612]: oh I did not see that

**Review by WaffleLapkin — APPROVED** [review]
LGTM, thanks!
(I've committed some small fmt changes so that I don't bother you with them)

**Pull request merged**